project_euler/problem_059/sol1.py

```diff
@@ -0,0 +1,121 @@
+method is using a sufficiently long password key for security, but short enough to
+be memorable.
+
+Your task has been made easy, as the encryption key consists of three lower case
+characters. Using p059_cipher.txt (right click and 'Save Link/Target As...'), a
+file containing the encrypted ASCII codes, and the knowledge that the plain text
+must contain common English words, decrypt the message and find the sum of the ASCII
+values in the original text.
+"""
+
+
+import os.path
+import string
+from itertools import cycle, product
+from typing import List, Optional, Set, Tuple
+
+VALID_CHARS: str = (
+    string.ascii_letters + string.digits + string.punctuation + string.whitespace
+)
+LOWERCASE_INTS: List[int] = list(map(ord, string.ascii_lowercase))
+VALID_INTS: Set[int] = set(map(ord, VALID_CHARS))
+
+COMMON_WORDS = ["the", "be", "to", "of", "and", "in", "that", "have"]
+
+
+def test_three_characters(ciphertext: List[int], key: Tuple[int, ...]) -> Optional[str]:
+    """
+    Given an encrypted message and a possible 3-character key, decrypt the message.
+    If the decrypted message contains a invalid character, i.e. not an ASCII letter,
+    a digit, punctuation or whitespace, then we know the key is incorrect, so return
+    None.
+    >>> test_three_characters([0, 17, 20, 4, 27], [104, 116, 120])
+    'hello'
+    >>> test_three_characters([68, 10, 300, 4, 27], [104, 116, 120]) is None
+    True
+    """
+    decoded: str = ""
+    keychar: int
+    cipherchar: int
+    decodedchar: int
+
+    for keychar, cipherchar in zip(cycle(key), ciphertext):
+        decodedchar = cipherchar ^ keychar
+        if decodedchar not in VALID_INTS:
+            return None
+        decoded += chr(decodedchar)
+
+    return decoded
+
+
+def filter_valid_chars(ciphertext: List[int]) -> List[str]:
+    """
+    Given an encrypted message, test all 3-character strings to try and find the
+    key. Return a list of the possible decrypted messages.
+    """
+    possibles: List[str] = []
+    for key in product(LOWERCASE_INTS, repeat=3):
+        encoded = test_three_characters(ciphertext, key)
+        if encoded is not None:
+            possibles.append(encoded)
+    return possibles
+
+
+def filter_common_word(possibles: List[str], common_word: str) -> List[str]:
+    """
+    Given a list of possible decoded messages, narrow down the possibilities
+    for checking for the presence of a specified common word. Only decoded messages
+    containing common_word will be returned.
+    """
+    return [possible for possible in possibles if common_word in possible]
+
+
+def solution() -> int:
+    """
+    Test the ciphertext against all possible 3-character keys, then narrow down the
+    possibilities by filtering using common words until there's only one possible
+    decoded message.
+    """
+    data: str
+    script_dir: str = os.path.abspath(os.path.dirname(__file__))
+    cipher_file: str = os.path.join(script_dir, "p059_cipher.txt")
```

Comment:
```suggestion
    data: str = Path(__file__).parent.joinpath("p059_cipher.txt").read_text(encoding="utf-8")
```
Please import: `from pathlib import Path`

Reply:
Thanks, didn't know that!